File: benchmarks/live_suite/kahan_dot.py
def run_kahan_dot(n):
    a = [(i % 100) + 1 for i in range(n)]
    b = [(i % 50) + 1 for i in range(n)]

    total_sum = 0
    comp = 0
    for j in range(n):
        term = a[j] * b[j]
        y = term - comp
        t = total_sum + y
        comp = (t - total_sum) - y
        total_sum = t

    return total_sum

if __name__ == "__main__":
    res = 0
    for _ in range(2):
        res = run_kahan_dot(500000)
    print(f"Kahan Dot Sum: {res}")

File: bench/live_suite/kahan_dot.rs
fn run_kahan_dot(n: usize) -> i64 {
    let mut a = Vec::with_capacity(n);
    let mut b = Vec::with_capacity(n);
    for i in 0..n {
        a.push(((i % 100) + 1) as i64);
        b.push(((i % 50) + 1) as i64);
    }

    let mut sum = 0i64;
    let mut comp = 0i64;
    for j in 0..n {
        let term = a[j] * b[j];
        let y = term - comp;
        let t = sum + y;
        comp = (t - sum) - y;
        sum = t;
    }

    sum
}

fn main() {
    let mut res = 0;
    for _ in 0..5 {
        res = run_kahan_dot(1000000);
    }
    println!("Kahan Dot Sum: {}", res);
}
